### PR TITLE
Feature/add bot wallet balance alarm

### PR DESCRIPTION
**Pull request by krisoshea-eth**
feat: adds bot wallet balance alarm 

**Motivation**

This PR was created in order to implement a bot wallet balance alarm in the bot script. This alarm will notify the team if the bot wallet balance falls below the predefined healthy threshold which is calculated using the following formula:  Wallet balance = minSponsorTokens * Number of open positions.

**Summary**

In order to achieve this, changes were made to the bot's index.js file which is run during the bot start-up process. The changes made to the index.js file allow for the bot wallet balance alarm to be created when the bot script is run.

**Testing**

Check a box to describe how you tested these changes and list the steps for reviewers to test.

- [ X]  Ran end-to-end test, running the code as in production

### CODE REVIEW

File: packages/liquidator/index.js [hunk 1]
```diff
@@ -155,6 +155,57 @@ async function run({
...
 
+    // Checks whether the liquidator bot wallet balance is >= total open positions * minimum sponsor amount
+    const calculateBotWalletBalance = async () => {
+      try {
+        // update client state
+        this.client = financialContractClient;
+        this.web3 = this.client.web3;
+        this.syntheticToken = syntheticToken;
+
+        await this.client.update();
```
Comment (KohliSuraj) [inline packages/liquidator/index.js]: we don't really need to update `financialContractClient` here again. Since, liquidator.update updates the financialContractClient too.

File: packages/liquidator/index.js [hunk 2]
```diff
@@ -155,6 +155,57 @@ async function run({
...
+
+        let minSponsorTokens = null;
+        if (liquidatorConfig.contractType === "ExpiringMultiParty") {
+          minSponsorTokens = await financialContract.methods.minSponsorTokens().call();
+        }
+
+        const numOfOpenPositions = this.client.getAllPositions().length;
+
+        console.log(`minSponsorTokens: ${minSponsorTokens}`);
```
Comment (KohliSuraj) [inline packages/liquidator/index.js]: I think we can use the variables already declared and fetched by this script previously, instead of fetching again?

**Review by KohliSuraj — CHANGES_REQUESTED** [review]

File: packages/disputer/index.js [hunk 3]
```diff
@@ -115,10 +116,8 @@ async function run({
 
     // Setup contract instances.
     const { getAbi: getVersionedAbi } = require(getContractsNodePackageAliasForVerion(disputerConfig.contractVersion));
-    const financialContract = new web3.eth.Contract(
-      getVersionedAbi(disputerConfig.contractType),
-      financialContractAddress
-    );
+    const abi = getVersionedAbi(disputerConfig.contractType);
```
Comment (KohliSuraj) [inline packages/disputer/index.js]: we can revert this change, seems unnecessary

File: packages/liquidator/src/walletBalanceAlarm.js [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+        message: "An error occurred during the calculation of the bot wallet balance",
+        error: error,
+      });
+
+      throw error;
+    }
+  }
+
+  async calculateTargetLiquidatorSynthBalance() {
+    this.numOfOpenPositions = await this.financialContractClient.getAllPositions().length;
```
Comment (KohliSuraj) [inline packages/liquidator/src/walletBalanceAlarm.js]: We should move `this.numOfOpenPositions = await this.financialContractClient.getAllPositions().length;` to the constructor, so that we can use this value throughout the class.

File: packages/liquidator/src/walletBalanceAlarm.js [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    return this.minSponsorTokens * this.numOfOpenPositions;
+  }
+
+  async calculateTargetLiquidatorCollateralBalance() {
+    const ooReward = await this.financialContract.methods.ooReward().call();
+    this.numOfOpenPositions = await this.financialContractClient.getAllPositions().length;
```
Comment (KohliSuraj) [inline packages/liquidator/src/walletBalanceAlarm.js]: reuse `this.numOfOpenPositions`

File: packages/liquidator/src/walletBalanceAlarm.js [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+        stack: error.stack,
+      });
+
+      throw error;
+    }
+  }
+
+  async calculateUsdcTargetBalance() {
+    const allPositions = await this.financialContractClient.getAllPositions();
+    this.totalCollateralAmount = await allPositions.reduce((total, position) => {
```
Comment (KohliSuraj) [inline packages/liquidator/src/walletBalanceAlarm.js]: I think we can directly use `totalPositionCollateral` from the contract, instead of recalculating here.

File: packages/liquidator/src/walletBalanceAlarm.js [hunk 7]
```diff
@@ -0,0 +1,157 @@
...
+    const ooReward = await this.financialContract.methods.ooReward().call();
+    console.log(`OO Reward: ${ooReward}`);
+
+    this.numOfOpenPositions = await this.financialContractClient.getAllPositions().length;
+    console.log(`Number of Open Positions: ${this.numOfOpenPositions}`);
+
+    this.disputerBondPercentage = await this.financialContract.methods.disputeBondPercentage().call();
+    console.log(`Dispute Bond Percentage: ${this.disputerBondPercentage / 1e18}`);
+
+    const usdcTargetBalance =
```
Comment (KohliSuraj) [inline packages/liquidator/src/walletBalanceAlarm.js]: This looks correct ✅

Comment (KohliSuraj) [reply packages/liquidator/src/walletBalanceAlarm.js]: I would just add extra brackets, to make sure correct order of calculation is being followed.

(this.totalCollateralAmount * (this.disputerBondPercentage / 1e18)) + ((this.numOfOpenPositions * this.bufferPercentage) * ooReward);

File: packages/liquidator/src/walletBalanceAlarm.js
Comment (KohliSuraj) [inline packages/liquidator/src/walletBalanceAlarm.js:1]: @krisoshea-eth Can you move the disputer logic to a separate file under **disputer/src** similar to what we do for liquidator? It'll be much more readable and cleaner.

Probably, I didn't mention this explicitly before, but I wanted you to create a new class for disputer similar to walletBalanceAlarm.js in liquidator.

**Review by KohliSuraj — CHANGES_REQUESTED** [review]

File: packages/liquidator/src/walletBalanceAlarm.js (hunk 4 above)
Comment (krisoshea-eth) [reply packages/liquidator/src/walletBalanceAlarm.js]: I don't think this is possible as we need to update the value of this.numOfOpenPositions regularly. If we move it to the constructor, it will always use the original value fetched in the constructor.

File: packages/liquidator/src/walletBalanceAlarm.js (hunk 6 above)
Comment (krisoshea-eth) [reply packages/liquidator/src/walletBalanceAlarm.js]: Good point! I didn't realise you could read that directly from the contract.

File: packages/liquidator/src/walletBalanceAlarm.js (hunk 4 above)
Comment (KohliSuraj) [reply packages/liquidator/src/walletBalanceAlarm.js]: @krisoshea-eth I agree with you 👍. Your approach is correct.

File: packages/liquidator/src/walletBalanceAlarm.js (hunk 7 above)
Comment (krisoshea-eth) [reply packages/liquidator/src/walletBalanceAlarm.js]: For some reason, everytime I push this change it reverts to the original version, even after I have disabled Prettier

**Comment by KohliSuraj** [comment]
Closing in favour of #31
